src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs

```diff
@@ -328,144 +340,568 @@ private void TrySetUnrecoverableError(SyntaxDiagnosticInfo error)
+                // inner three braces start the interpolation.  The outer two braces are just content.  This ensures the
+                // rule that the content cannot contain a sequence of open or close braces equal to (or longer than) the
+                // dollar sequence.
+                var beforeOpenBracesPosition = _lexer.TextWindow.Position;
+                var openBraceCount = _lexer.ConsumeOpenBraceSequence();
+                if (openBraceCount < startingDollarSignCount)
+                {
+                    // not enough open braces to matter.  Just treat as content.
+                    return;
+                }
+
+                var afterOpenBracePosition = _lexer.TextWindow.Position;
+                if (openBraceCount >= 2 * startingDollarSignCount)
+                {
+                    // Too many open braces.  Report an error on the portion up before the section that counts as the
+                    // start of the interpolation.
+                    TrySetError(_lexer.MakeError(
+                        beforeOpenBracesPosition,
+                        width: openBraceCount - startingDollarSignCount,
+                        ErrorCode.ERR_TooManyOpenBracesForRawString));
+                }
+
+                // Now, try to scan the contents of the interpolation.  Ending when we hit a close brace.
+                ScanInterpolatedStringLiteralHoleBalancedText(kind, '}', isHole: true, out var colonRange);
+
+                var beforeCloseBracePosition = _lexer.TextWindow.Position;
+                var closeBraceCount = _lexer.ConsumeCloseBraceSequence();
+
+                if (closeBraceCount == 0)
+                {
+                    // Didn't find any close braces.  Report a particular error on the open braces that they are unclosed.
+                    TrySetError(_lexer.MakeError(
+                        position: afterOpenBracePosition - startingDollarSignCount,
+                        width: startingDollarSignCount,
+                        ErrorCode.ERR_UnclosedExpressionHole));
+                }
+                else if (closeBraceCount < startingDollarSignCount)
+                {
+                    // not enough close braces to end the interpolation.  Report here.
+                    TrySetError(_lexer.MakeError(
+                        beforeOpenBracesPosition,
+                        width: openBraceCount - startingDollarSignCount,
+                        ErrorCode.ERR_NotEnoughCloseBracesForRawString));
+                }
+                else
+                {
+                    // Only consume up to the minimum number of close braces we need to end the interpolation. Any
+                    // excess will be consumed in the content consumption pass in ScanInterpolatedStringLiteralContents.
+                    _lexer.TextWindow.Reset(beforeCloseBracePosition + startingDollarSignCount);
+                }
+
+                interpolations?.Add(new Interpolation(
```

Comment:
Curious when this would ever be null. During rescanning?

Reply:
yes.  this is a core part of interpolated strings.  during *lexing* the interpolations-array is null.  lexing just wants to find the full span of the string-literal.  However, during parsing we actually fill this array so the parser know which sections of the literal are interpolations that it should then parse subexpressions out of.  i can doc.